### PR TITLE
Proposal: function to quickly generate `y` for decoding schemes

### PR DESCRIPTION
**Pull request by wmvanvliet**
Per request from Lauri Parkkonen, a proposal for a new function.

For decoding tasks, we need a function to conveniently construct the `y` variable for scikit-learn models. This PR adds a function `dummy_encoding` that tries to do the right thing when given an `Epochs` object, so you can do this:

    epochs = epochs[['left', 'right']]
    X = epochs.get_data()
    y = mne.decoding.dummy_encoding(epochs)

See the docstring of the function for details.

### CODE REVIEW

**Comment by agramfort** [comment]
I think it's a great idea. I am just not too found of the name.

It does basically what LabelEncoder does in sklearn:

http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html#sklearn.preprocessing.LabelEncoder

I would call it make_labels or ... ? any suggestion?

**Comment by jasmainak** [comment]
I think some of the GAT examples might benefit from this ...

**Comment by LauriParkkonen** [comment]
Thanks @wmvanvliet for this! I was originally thinking of something like
`y = epochs.get_categories()`
but that might be difficult to implement because that would mean changing the Epoch object. And since sklearn does not like string labels, those would still need to be mapped onto integers.
I agree with @agramfort that mne.decoding.make_labels() might be a good name :)

**Comment by LauriParkkonen** [comment]
Two further points that I've already discussed with @wmvanvliet.
1. Using the trigger codes as labels is not good idea since we frequently have situations where the classification categories are defined conditionally on the triggers codes, e.g., Cat1 comprises only those trials with trigger 1 that are preceded by trials with trigger code 2, and Cat2 is those trials with trigger 1 that are preceded with trials with trigger code 3. So, if we sort the epochs properly into such categories, all of them will have the same trigger code. Therefore, the category label should somehow make its way to the label of each of its epochs.
2. The order of the epochs in the raw data file should preserved. For typical off-line classification this may not matter because the samples are suffled anyway but for mimicing real-time decoding (training the classifier with samples from the beginning of the recording only) this does matter.

**Comment by kingjr** [comment]
I don't understand. Why don't you just do
`sklearn.preprocessing.MultiLabelBinarizer().fit_transform(epochs.events[:,
[2]])`?

Note that the definition of the independent variables is highly variable
across experimental setup (discriminatory vs generative, hierarchichal,
discrete vs continuous etc). IMO it would be better to clarify the example
or the documentation than adding this function. But perhaps I missed a
particular constrain.

**Comment by wmvanvliet** [comment]
@kingjr I should've known that scikit-learn would already implement something like this. Good, that means things will be simpler.

In my opinion, this:

    sklearn.preprocessing.MultiLabelBinarizer().fit_transform(epochs.events[:, [2]])

is good, but not great, because it requires the following steps of thought:

 1. The class/condition an epoch belongs to is stored in the `.events` property (not something a beginner knows)
 2. Specifically, it is stored in the last column of this matrix (definitely not something a beginner knows)
 3. When selecting epochs, the `.events` property is updated to reflect these changes, so it is safe to use (something you may easily deduce when playing around, but not something that is intuitively clear).

I would like there to be an easy interface to getting "class labels" that beginners, with no deep knowledge of the various properties of the Epoch data object, can find in the API docs.

**Comment by kingjr** [comment]
It seems to me that these are documentation/example issues, not calls for a
new function.

**Comment by jasmainak** [comment]
Indeed, we should not reinvent the wheel. An update to an example would be nice. But I think `MultiLabelBinarizer` does not address the issue of categories defined based on temporal relationships between triggers. Although this PR in its current form does not either ...

**Comment by wmvanvliet** [comment]
If we can make intuitive code examples without adding new functions, all the better.

File: mne/decoding/base.py [hunk 1]
```diff
@@ -746,3 +746,98 @@ def _make_scorer(scoring):
...
+    epoch belongs to. By default, the columns of the matrix will correspond to
+    the events in order of their respective event-ids (further ordered by their
+    event name if multiple classes share the same event-id). Use the
+    `use_events` parameter to specify a different order.
+
+    Parameters
+    ----------
+    epochs : instance of :class:`mne.Epochs`
+        The epochs to create the dummy encoding for.
+    use_events : list of str | None
```
Comment (jasmainak) [inline mne/decoding/base.py]: @kingjr this is what `MultilabelBinarizer` cannot do ...

Comment (wmvanvliet) [inline mne/decoding/base.py]: but admittedly, this can be done with:

    epochs[['a', 'b']]

Comment (jasmainak) [inline mne/decoding/base.py]: it's not the same because you lose the ordering if you do it, no? I am not sure ...

Comment (kingjr) [inline mne/decoding/base.py]: if this removes the trial order (which surprises me..), we should correct for that at the source.

**Comment by agramfort** [comment]
any consensus here?

**Comment by LauriParkkonen** [comment]
In the lab, we've had some further discussions on this with @wmvanvliet but we haven't been able to come up with THE way. Perhaps this should be considered in conjunction with conditional averaging, i.e., how to label/categorize epochs with more complex rules than just going for a particular trigger code (there is, of course, a limit what can be supported because one could come up with arbitrarily complex rules). Human-readable category names (labels) would be nice because the decoding plot functions could use them automatically. Well, no elegant suggestions yet...

**Comment by agramfort** [comment]
one way could be to have event_id be a callable such as the function

def event_id(events, k):
    if events[k] == 1 and events[k] == 2:
          return True
    else:
          return False

that would allow you to say what to include in the average. But it will
break some stuff such as IO and maybe more.

I am not sure we need to go complex like this.

Let's try to keep it simple and stupid :)

**Comment by wmvanvliet** [comment]
How about we just add a new property to the `Epochs` object:

```python
def get_labels(self):
    return self.events[:, 2]

def set_labels(self, labels, conditions=None):
    if conditions is not None:
        self.event_id = conditions
    self.events[:, 2] = labels
```

Then, in the vast majority of cases, this is all you need

```python
epochs = epochs[['first_class', 'second_class']]
X = epochs.get_data()
y = epochs.get_labels()
```

In the above case, the labels are based on the trigger codes. If you want the labels to be based on something else, you can do this:

```python
# Construct labels any way you want. Use epochs.events and epochs.get_labels() as needed.
new_labels = ...
# Also supply how labels map to conditions for plotting functions
epochs.set_labels(new_labels, dict(first_class=-1, second_class=1))
```

**Comment by wmvanvliet** [comment]
All the above does is implement a purely cosmetic wrapper around the functionality we already have. The  reason is to make your scripts more readable and have the example scripts more intuitive. IMO, for MNE-Python, this reason is important enough to warrant adding such a cosmetic wrapper.

**Comment by jona-sassenhagen** [comment]
The epochs method idea is nice.

Maybe the ability to set labels is overkill for now. Why not for now just have a get_labels method that can take an optional conditions_to_labels dict?

**Comment by agramfort** [comment]
>
> Maybe the ability to set labels is overkill for now. Why not for now just
> have a get_labels method that can take an optional conditions_to_labels
> dict?


+1

**Comment by wmvanvliet** [comment]
> Why not for now just have a get_labels method that can take an optional conditions_to_labels dict?

What would passing a conditions_to_labels dict accomplish in this case?

**Comment by agramfort** [comment]
convert eg auditory to 1 or visual to 2.

event_id = {'Auditory/Left': 1, 'Auditory/Right': 2,
            'Visual/Left': 3, 'Visual/Right': 4}

epochs = ...

y = epochs.get_label({'Auditory':-1, 'Visual':1}

**Comment by wmvanvliet** [comment]
That's an excellent idea for the `get_labels` function, but I still would like a `set_labels` function  to match.
The use case for `set_labels` is to make it easy to assign epochs to conditions, based on something else then the trigger codes. Some examples:

 1. During the measurement, we also recorded reaction time. Now I want to create evokeds based on quick versus slow responses.
 2. My conditions depend on a button press during the previous trial.
 3. I'm working with data provided by a third party. The trigger codes don't make much sense and I want to assign my own conditions, based on parsing the logfile.
 4. During analysis, I think of a new contrast to try, now I want to quickly create new evokeds based on the contrast I just thought of.

**Comment by Eric89GXL** [comment]
It seems like the most flexible approach would be something like:
```
def set_event_ids(event_id, event_kinds=None)
    """Set new event IDs.

    Parameters
    ----------
    event_id : dict
        The new event dict.
    event_kinds : ndarray, shape (n_events,)
        The integer IDs for the new events. If None, ``epochs.events[:, 2]`` is used.
        Must have the same length as the existing ``epochs.events``.
    ```

**Comment by Eric89GXL** [comment]
> 1. During the measurement, we also recorded reaction time. Now I want to create evokeds based on quick versus slow responses.
> 2. My conditions depend on a button press during the previous trial.
> 3. I'm working with data provided by a third party. The trigger codes don't make much sense and I want to assign my own conditions, based on parsing the logfile.

FWIW all of these use cases are currently supported by changing `events` before creating epochs, so hopefully you haven't been slowed by them too much. But I agree having a way to do it after the fact, especially for case (4) above, would be worth it. I think the code I pasted above should take care of all of these use cases. I don't think `labels` is the correct term, though, since at the MNE end we're really talking about how to recode `event_id`s, as far as I can tell. Allowing flexibility with `event_id` and `events` in MNE should immediately give the functionality you need to create `labels` in `sklearn`, hopefully.

**Comment by kingjr** [comment]
I am personally skeptical. The main issues that I foresee with a `set_label` function or a new API is that 1) it will not take into account continuous values (required for regression problems (including your reaction time example)) and 2) it suggests the user to change the labels back and forth, rather than keeping a single "conditions" array - which is arguably less dangerous.

In an ideal world, the conditions wouldn't have been stored in a strict n_sample * 3 integer array, but apparently we can't change that.

Consequently,  I would recommend teaching advanced users, who want to keep track of multiple/hierarchical/continuous labels to use and pandas dataframe on the side: e.g.

```python
sel = conditions.query('reaction_times == 'slow' and modality == 'auditory').index
evoked = epochs[sel].average()
```

**Comment by agramfort** [comment]
I agree with @kingjr

**Comment by wmvanvliet** [comment]
Closing this as the consensus is that the way to go forward is to improve the doc, not add more methods/functions.
